File: app/src/main/java/jaanc/cents/ui/expenselistpage/ExpenseListTile.kt
package jaanc.cents.widgets

import androidx.compose.foundation.layout.*
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Surface
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import jaanc.cents.domain.Amount
import jaanc.cents.domain.Expense
import jaanc.cents.domain.ExpenseCategory
import jaanc.cents.utils.display
import jaanc.cents.utils.displayRelative

@Composable
fun ExpenseListTile(expense: Expense) {
    val typography = MaterialTheme.typography
    val colors = MaterialTheme.colors

    Surface() {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(
                expense.createdAt.displayRelative(), style = typography.overline
            )

            Row(
                horizontalArrangement = Arrangement.SpaceBetween,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(
                    if (expense.category.name.isNotBlank()) expense.category.name
                    else "Uncategorized",
                    style = typography.subtitle1,
                    modifier = Modifier.alignByBaseline()
                )
                Text(
                    expense.cost.display(),
                    style = typography.subtitle1,
                    modifier = Modifier.alignByBaseline()
                )
            }

            if (expense.note.isNotBlank()) {
                Text(
                    expense.note,
                    style = typography.subtitle2,
                    color = colors.onSurface.copy(alpha = 0.6f),
                    maxLines = 3,
                    overflow = TextOverflow.Ellipsis
                )
            }
        }
    }
}

@Preview
@Composable
fun ExpenseListTilePreview() {
    ExpenseListTile(
        Expense(
            category = ExpenseCategory("Food"),
            cost = Amount.from(12),
            note = "Cofe!"
        )
    )
}

File: lib/src/ui/expense_list_page/expense_list_tile.dart
import 'package:cents/src/domain/expense.dart';
import 'package:cents/src/domain/ext_date_time.dart';
import 'package:cents/src/domain/ext_string.dart';
import 'package:flutter/material.dart';

class ExpenseListTile extends StatelessWidget {
  final Expense expense;
  final VoidCallback onTap;
  final VoidCallback? onLongPress;
  final bool isSelected;

  const ExpenseListTile({super.key, 
    required this.expense,
    required this.onTap,
    this.onLongPress,
    required this.isSelected,
  });

  @override
  Widget build(BuildContext context) {
    final colorScheme = Theme.of(context).colorScheme;
    final selectColor = colorScheme.onSurface.withAlpha(33);

    return Material(
      color: isSelected ? selectColor : null,
      child: InkWell(
        onTap: onTap,
        onLongPress: onLongPress,
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              _createdAtText(context),
              Row(
                mainAxisSize: MainAxisSize.max,
                crossAxisAlignment: CrossAxisAlignment.baseline,
                textBaseline: TextBaseline.alphabetic,
                children: [
                  Expanded(child: _categoryText(context)),
                  _costText(context),
                ],
              ),
              if (expense.note.isNotBlank) _noteText(context),
            ],
          ),
        ),
      ),
    );
  }

  Widget _createdAtText(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;

    return Text(expense.createdAt.relativeDisplay(), style: textTheme.labelSmall);
  }

  Widget _categoryText(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;

    return Text(
      expense.category.name.isNotBlank
          ? expense.category.name
          : 'Uncategorized',
      style: textTheme.titleMedium,
      maxLines: 1,
      overflow: TextOverflow.ellipsis,
    );
  }

  Widget _costText(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;

    return Text(expense.cost.toLocalString(), style: textTheme.titleMedium);
  }

  Widget _noteText(BuildContext context) {
    final theme = Theme.of(context);
    final textTheme = theme.textTheme;
    final brightness = theme.brightness;
    final textColor =
        brightness == Brightness.light ? Colors.black54 : Colors.white54;

    return Text(
      expense.note,
      style: textTheme.bodyMedium?.apply(color: textColor),
    );
  }
}
